syft/internal/cache/resolver.go

```diff
@@ -0,0 +1,145 @@
+package cache
+
+import (
+	"bytes"
+	"encoding/json"
+	"fmt"
+	"io"
+	"net/url"
+	"regexp"
+	"strings"
+	"time"
+
+	"github.com/anchore/syft/internal"
+	"github.com/anchore/syft/internal/log"
+	"github.com/anchore/syft/syft/file"
+	"github.com/anchore/syft/syft/internal/fileresolver"
+)
+
+// Resolver interface provides a single Resolve method, which will return from cache
+// or call the provided resolve function to get the value if not available in cache
+type Resolver[T any] interface {
+	Resolve(key string, resolver resolverFunc[T]) (T, error)
+}
+
+// GetResolver returns a cache resolver for persistent cached data across Syft runs
+func GetResolver[T any](name string) Resolver[T] {
+	versionKey := hashType[T]()
+	return cacheResolver[T]{
+		name:         fmt.Sprintf("%s/%s", name, versionKey),
+		ttl:          7 * 24 * time.Hour,
+		fileResolver: fileResolver(name, versionKey),
```

Comment:
[no request for change in this PR] the reason for using a `file.Resolver` earlier was because we were in the context of the catalogers, which opted to not deal with filesystem calls directly.

We can leave this for now, but it might be in the future we use `afero.FS` directly here (or `fs.FS` if we decide we don't need symlink support).

Reply:
Well, I take this back (or it's more grey than I thought) -- we might cache entire dir trees that should be used as part of cataloging, and expose that resolver out to the cataloger. Which means catalogers only needing to reason about one way to get to files is a good thing (via the resolvers).

Reply:
I've changed this so the cache is basically a simple key/[]byte store. if a user wanted to store a directory tree, it would be possible to use the cache directly and store, say an uncompressed .tar file. it returns a ReaderAt, so it could be efficient to traverse. we could add some cache.DirTree() utility of some sort to help this if we really needed/wanted it, but I don't really think we want this.